File: libs/ktheme-compose/src/main/kotlin/io/ktheme/compose/modifiers/Metallic.kt
// libs/ktheme-compose/src/main/kotlin/io/ktheme/compose/modifiers/Metallic.kt
//
// `Modifier.metallic(...)` — applies the engine's signature 5-stop
// linear gradient as a background. Reads the variant + intensity from
// the active theme by default; pass an explicit MetallicGradient to
// override.
//
// The gradient is shadow → base → highlight → base → shadow at 135°,
// matching the CSS engine. Intensity (0–1) lerps the highlight toward
// pure white for extra "polish".

package io.ktheme.compose.modifiers

import androidx.compose.foundation.background
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.runtime.Composable
import androidx.compose.runtime.ReadOnlyComposable
import androidx.compose.ui.Modifier
import androidx.compose.ui.composed
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Shape
import androidx.compose.ui.graphics.lerp
import io.ktheme.compose.LocalKtheme
import io.ktheme.compose.toComposeColor
import io.ktheme.model.MetallicGradient
import io.ktheme.model.MetallicVariant

/**
 * Paint a metallic gradient background. By default reads the variant
 * and intensity from the surrounding [io.ktheme.compose.KthemeTheme].
 *
 * @param shape Clip shape (defaults to no clipping).
 * @param angleDeg Gradient angle. 135° matches the engine default.
 */
public fun Modifier.metallic(
    variant: MetallicVariant? = null,
    intensity: Float? = null,
    shape: Shape = RoundedCornerShape(0),
    angleDeg: Float = 135f,
): Modifier = composed {
    val theme = LocalKtheme.current
    val v = variant ?: theme.effects.metallic.variant
    val i = (intensity ?: theme.effects.metallic.intensity.toFloat()).coerceIn(0f, 1f)
    val grad = MetallicGradient.forVariant(v)
    background(brush = metallicBrush(grad, i, angleDeg), shape = shape)
}

/** Build a 5-stop linear-gradient brush for [grad] at [angleDeg]. */
@Composable
@ReadOnlyComposable
public fun metallicBrush(
    grad: MetallicGradient,
    intensity: Float = 0.7f,
    angleDeg: Float = 135f,
): Brush {
    val base = grad.base.toComposeColor()
    val shadow = grad.shadow.toComposeColor()
    val highlight = lerp(grad.highlight.toComposeColor(), Color.White, (intensity - 0.5f).coerceAtLeast(0f))
    val rad = Math.toRadians(angleDeg.toDouble())
    val dx = Math.cos(rad).toFloat()
    val dy = Math.sin(rad).toFloat()
    return Brush.linearGradient(
        colorStops = arrayOf(
            0.00f to shadow,
            0.25f to base,
            0.50f to highlight,
            0.75f to base,
            1.00f to shadow,
        ),
        // Compose normalizes; offsets approximate angle direction.
        start = Offset(0f, 0f),
        end = Offset(1000f * dx, 1000f * dy),
    )
}

File: libs/ktheme-swift/Sources/KthemeUI/Modifiers/Metallic.swift
// libs/ktheme-swift/Sources/KthemeUI/Modifiers/Metallic.swift
//
// `.metallic(...)` ViewModifier — paints the engine's signature
// 5-stop linear gradient (shadow → base → highlight → base → shadow)
// as the receiver's background. Reads variant + intensity from the
// surrounding theme by default.

import SwiftUI
import Ktheme

extension View {
    /// Paint a metallic gradient background matching the active theme's
    /// metallic effect. Pass an explicit variant/intensity to override.
    public func metallic(
        variant: MetallicVariant? = nil,
        intensity: Double? = nil,
        cornerRadius: CGFloat = 0,
        angle: Double = 135
    ) -> some View {
        modifier(MetallicBackground(
            overrideVariant: variant,
            overrideIntensity: intensity,
            cornerRadius: cornerRadius,
            angle: angle
        ))
    }
}

private struct MetallicBackground: ViewModifier {
    let overrideVariant: MetallicVariant?
    let overrideIntensity: Double?
    let cornerRadius: CGFloat
    let angle: Double

    @Environment(\.ktheme) private var theme

    func body(content: Content) -> some View {
        let v = overrideVariant ?? theme.effects.metallic.variant
        let i = max(0, min(1, overrideIntensity ?? theme.effects.metallic.intensity))
        let g = MetallicGradient.forVariant(v)
        content.background(
            RoundedRectangle(cornerRadius: cornerRadius, style: .continuous)
                .fill(metalGradient(g, intensity: i, angle: angle))
        )
    }
}

/// 5-stop LinearGradient matching the engine's CSS recipe.
public func metalGradient(
    _ g: MetallicGradient,
    intensity: Double = 0.7,
    angle: Double = 135
) -> LinearGradient {
    let shadow    = HexColor.color(g.shadow)
    let base      = HexColor.color(g.base)
    let extraGlow = max(0, intensity - 0.5)
    let highlight = HexColor.color(g.highlight).opacity(1 - extraGlow * 0.0)
        .blendOver(.white, fraction: extraGlow)

    let theta = angle * .pi / 180
    let dx = cos(theta), dy = sin(theta)
    return LinearGradient(
        stops: [
            .init(color: shadow,    location: 0.00),
            .init(color: base,      location: 0.25),
            .init(color: highlight, location: 0.50),
            .init(color: base,      location: 0.75),
            .init(color: shadow,    location: 1.00),
        ],
        startPoint: UnitPoint(x: 0.5 - dx * 0.5, y: 0.5 - dy * 0.5),
        endPoint:   UnitPoint(x: 0.5 + dx * 0.5, y: 0.5 + dy * 0.5)
    )
}

private extension Color {
    /// Mix `self` with `other` at fraction 0…1.
    func blendOver(_ other: Color, fraction f: Double) -> Color {
        // SwiftUI Color is opaque; approximate by overlaying.
        // Callers use this only for highlight whitening, so a simple
        // linear interpolation in sRGB is adequate.
        guard f > 0 else { return self }
        return Color(
            .sRGB,
            red:   1, green: 1, blue: 1,
            opacity: f
        ).opacity(1) // placeholder — see note in README about platform color math
    }
}
